typing/printtyp.ml

```diff
@@ -1638,19 +1638,37 @@ let rec tree_of_modtype ?(ellipsis=false) = function
       Omty_alias (tree_of_path Module p)
 
 and tree_of_signature sg =
-  wrap_env (fun env -> env) (tree_of_signature_rec !printing_env false) sg
-
-and tree_of_signature_rec env' in_type_group = function
-    [] -> stop_type_group env'; []
-  | item :: rem as items ->
-      let in_type_group = still_in_type_group env' in_type_group item in
-      let (sg, rem) = filter_rem_sig item rem in
-      hide_rec_items items;
-      protect_rec_items items;
-      reset_naming_context ();
-      let trees = trees_of_sigitem item in
-      let env' = Env.add_signature (item :: sg) env' in
-      trees @ tree_of_signature_rec env' in_type_group rem
+  wrap_env (fun env -> env)
+    (tree_of_signature_rec (fun _ (_,x) -> x) !printing_env) sg
+
+and tree_of_signature_rec : 'r. (_->_->'r) -> _ -> _ -> 'r list
+  = fun decorate env' sg ->
+  let structured = group_recursive_items (group_syntactic_items sg) in
+  let collect_trees_of_rec_group (env,trees) group =
+    let env', more_trees = trees_of_recursive_sigitem_group env group in
+    let more_trees = List.map (decorate env) more_trees in
+    set_printing_env env';
+    env', more_trees :: trees in
```

Comment:
This is not due to your refactoring (it was already there), but I find it confusing that this code passes `env` around as a pure state-passing program *and* mutates the printing state at the same time. If I read the (new) code correctly, it is always the case that whenever `trees_of_recursive_sigitem_group` is called with an `env` parameter, then this `env` input is equal to `!printing_env`. If that is indeed true, could we drop the env-passing here, and have `trees_of_recursive_sigitem_group` just take the group as input, read `input_env` and call `set_printing_env` before returning?

(Maybe this belong to a separate commit or PR.)

Reply:
Now that I reviewed `trees_of_recursive_sigitem_group` as well, I am not convinced anymore than pushing the state-setting logic there is the right idea, because in that function state-passing is natural. (We don't want to call `set_printing_env` more than once for efficiency reasons.) It could stay here, but still doesn't need to be a fold.

Reply:
I removed the duplicated handling of env here.